compiler/nilcheck.nim

```diff
@@ -0,0 +1,667 @@
+# map is env?
+# a or b
+#   each one forks a different env
+#   result = union(envL, envR)
+# a and b
+#   b forks a's env
+# if a: code
+#   result = union(previousEnv after not a, env after code)
+# if a: b else: c
+#   result = union(env after b, env after c)
+# result = b
+#   nilability result <- nilability b, if return type is not nil and result not safe, error
+# return b
+#   as result = b
+# try: a catch: b finally: c
+#   in b and c env is union of all possible try first n lines, after union of a and b and c
+# case a: of b: c
+#   similar to if
+# call(arg)
+#   if it returns ref, assume it's MaybeNil: hint that one can add not nil to the return type
+# call(var arg) # zahary comment
+#   if arg is ref, assume it's MaybeNil after call
+# loop
+#   union of env for 0, 1, 2 iterations as Herb Sutter's paper
+# fields and index
+#   basically we generate a id for them and do almost the same as
+#   for names, we preserve nilability for them
+#   however, we also maintain that their atom variables are
+#   their dependencies
+#   and when the atom variables change, they invalidate
+#   the compound expressions and they are MaybeNil
+# return
+#   if something: stop (break return etc)
+#   is equivalent to if something: .. else: remain
+# new(ref)
+#   ref becomes Safe
+# each check returns its nilability and map
+
+type
+  Symbol = int
+
+  TransitionKind = enum TArg, TAssign, TType, TNil, TVarArg, TResult, TOwner
+  
+  History = object
+    info: TLineInfo
+    nilability: Nilability
+    kind: TransitionKind
+    
+
+  NilMap* = ref object
+    locals*:   Table[Symbol, Nilability]
+    history*:  Table[Symbol, seq[History]]
+    previous*: NilMap
+    base*:     NilMap
+    top*:      NilMap
+    dependencies*: Table[Symbol, seq[Symbol]]
+
+  Nilability* = enum Safe, MaybeNil, Nil
+
+  Check = tuple[nilability: Nilability, map: NilMap]
+
+let RESULT_ID = -1
+
+proc check(n: PNode, conf: ConfigRef, map: NilMap): Check
+proc checkCondition(n: PNode, conf: ConfigRef, map: NilMap, isElse: bool, base: bool): NilMap
+proc invalidate(map: NilMap, name: Symbol, info: TLineInfo)
+
+proc newNilMap(previous: NilMap = nil, base: NilMap = nil): NilMap =
+  result = NilMap(
+    locals: initTable[Symbol, Nilability](),
+    previous: previous,
+    base: base,
+    dependencies: initTable[Symbol, seq[Symbol]](),
+    history: initTable[Symbol, seq[History]]())
+  result.top = if previous.isNil: result else: previous.top
+
+proc `[]`(map: NilMap, name: Symbol): Nilability =
+  var now = map
+  while not now.isNil:
+    if now.locals.hasKey(name):
+      return now.locals[name]
+    now = now.previous
+  return Safe
+
+proc history(map: NilMap, name: Symbol): seq[History] =
+  var now = map
+  var h: seq[History] = @[]
+  while not now.isNil:
+    if now.history.hasKey(name):
+      # h = h.concat(now.history[name])
+      return now.history[name]
+    now = now.previous
+  return @[]
+
+import macros
+
+macro aecho*(a: varargs[untyped]): untyped =
```

Comment:
wouldn't a simpler definition work here:

```nim
template trace(args: varargs[untyped]) =
   when debugInfo:
      echo args
```

Reply:
yes, I agree, maybe I didn't know that varargs work in this situation